CamelSources not working in nightly

Get this error:

```
Exception in thread "main" org.apache.camel.RuntimeCamelException: java.lang.RuntimeException: Error creating instance of loader: knative-source
	at org.apache.camel.RuntimeCamelException.wrapRuntimeCamelException(RuntimeCamelException.java:52)
	at org.apache.camel.k.listener.RoutesConfigurer.load(RoutesConfigurer.java:64)
	at org.apache.camel.k.listener.RoutesConfigurer.accept(RoutesConfigurer.java:52)
	at org.apache.camel.k.listener.AbstractPhaseListener.accept(AbstractPhaseListener.java:32)
	at org.apache.camel.k.main.ApplicationRuntime$MainListenerAdapter.lambda$invokeListeners$0(ApplicationRuntime.java:179)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.SortedOps$SizedRefSortingSink.end(SortedOps.java:357)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at org.apache.camel.k.main.ApplicationRuntime$MainListenerAdapter.invokeListeners(ApplicationRuntime.java:178)
	at org.apache.camel.k.main.ApplicationRuntime$MainListenerAdapter.beforeConfigure(ApplicationRuntime.java:152)
	at org.apache.camel.main.BaseMainSupport.postProcessCamelContext(BaseMainSupport.java:584)
	at org.apache.camel.main.BaseMainSupport.initCamelContext(BaseMainSupport.java:423)
	at org.apache.camel.k.main.ApplicationRuntime$MainAdapter.doInit(ApplicationRuntime.java:195)
	at org.apache.camel.support.service.ServiceSupport.init(ServiceSupport.java:83)
	at org.apache.camel.support.service.ServiceSupport.start(ServiceSupport.java:112)
	at org.apache.camel.main.MainSupport.run(MainSupport.java:82)
	at org.apache.camel.k.main.ApplicationRuntime.run(ApplicationRuntime.java:71)
	at org.apache.camel.k.main.Application.main(Application.java:42)
Caused by: java.lang.RuntimeException: Error creating instance of loader: knative-source
	at org.apache.camel.k.support.RuntimeSupport.lambda$lookupLoaderFromResource$10(RuntimeSupport.java:212)
	at java.base/java.util.Optional.orElseThrow(Optional.java:408)
	at org.apache.camel.k.support.RuntimeSupport.lookupLoaderFromResource(RuntimeSupport.java:212)
	at org.apache.camel.k.support.RuntimeSupport.lookupLoaderById(RuntimeSupport.java:189)
	at org.apache.camel.k.support.RuntimeSupport.lambda$loaderFor$8(RuntimeSupport.java:173)
	at java.base/java.util.Optional.map(Optional.java:265)
	at org.apache.camel.k.support.RuntimeSupport.loaderFor(RuntimeSupport.java:172)
	at org.apache.camel.k.listener.RoutesConfigurer.load(RoutesConfigurer.java:82)
	at org.apache.camel.k.listener.RoutesConfigurer.load(RoutesConfigurer.java:62)
	... 21 more
```


Apparently the dependencies for a simple timer to channel contain:
```
    - camel:timer
    - mvn:org.apache.camel.k/camel-k-runtime-main
```

While in RC2 it was:
```
    - camel:timer
    - mvn:org.apache.camel.k/camel-k-loader-knative
    - mvn:org.apache.camel.k/camel-k-loader-yaml
    - mvn:org.apache.camel.k/camel-k-runtime-knative
    - mvn:org.apache.camel.k/camel-k-runtime-main
```

We can add an example of Camel Source from last released Knative in the CI to avoid such regressions.

--- a/pkg/controller/integration/initialize.go
+++ b/pkg/controller/integration/initialize.go
@@ -21,11 +21,8 @@ import (
 	"context"
 
 	v1 "github.com/apache/camel-k/pkg/apis/camel/v1"
-	"github.com/apache/camel-k/pkg/client"
-	"github.com/apache/camel-k/pkg/platform"
 	"github.com/apache/camel-k/pkg/trait"
 	"github.com/apache/camel-k/pkg/util/defaults"
-	"github.com/apache/camel-k/pkg/util/knative"
 )
 
 // NewInitializeAction creates a new initialize action
@@ -53,36 +50,11 @@ func (action *initializeAction) Handle(ctx context.Context, integration *v1.Inte
 		return nil, err
 	}
 
-	pl, err := platform.GetCurrentPlatform(ctx, action.client, integration.Namespace)
-	if err != nil {
-		return nil, err
-	}
-
 	kit := v1.NewIntegrationKit(integration.Namespace, integration.Spec.Kit)
 
 	integration.Status.Phase = v1.IntegrationPhaseBuildingKit
 	integration.SetIntegrationKit(&kit)
-	integration.Status.Profile = determineBestProfile(ctx, action.client, integration, pl)
 	integration.Status.Version = defaults.Version
 
 	return integration, nil
-}
-
-// DetermineBestProfile tries to detect the best trait profile for the integration
-func determineBestProfile(ctx context.Context, c client.Client, integration *v1.Integration, p *v1.IntegrationPlatform) v1.TraitProfile {
-	if integration.Spec.Profile != "" {
-		return integration.Spec.Profile
-	}
-	if integration.Status.Profile != "" {
-		// Integration already has a profile
-		return integration.Status.Profile
-	}
-	if p.Status.Profile != "" {
-		// Use platform profile if set
-		return p.Status.Profile
-	}
-	if knative.IsEnabledInNamespace(ctx, c, p.Namespace) {
-		return v1.TraitProfileKnative
-	}
-	return platform.GetProfile(p)
 }

--- a/pkg/controller/integration/platform_setup.go
+++ b/pkg/controller/integration/platform_setup.go
@@ -20,6 +20,11 @@ package integration
 import (
 	"context"
 
+	"github.com/apache/camel-k/pkg/client"
+	"github.com/apache/camel-k/pkg/platform"
+	"github.com/apache/camel-k/pkg/util/knative"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+
 	v1 "github.com/apache/camel-k/pkg/apis/camel/v1"
 	"github.com/apache/camel-k/pkg/trait"
 )
@@ -50,5 +55,31 @@ func (action *platformSetupAction) Handle(ctx context.Context, integration *v1.I
 		return nil, err
 	}
 
+	pl, err := platform.GetCurrentPlatform(ctx, action.client, integration.Namespace)
+	if err != nil && !k8serrors.IsNotFound(err) {
+		return nil, err
+	} else if pl != nil {
+		integration.Status.Profile = determineBestProfile(ctx, action.client, integration, pl)
+	}
+
 	return integration, nil
+}
+
+// DetermineBestProfile tries to detect the best trait profile for the integration
+func determineBestProfile(ctx context.Context, c client.Client, integration *v1.Integration, p *v1.IntegrationPlatform) v1.TraitProfile {
+	if integration.Spec.Profile != "" {
+		return integration.Spec.Profile
+	}
+	if integration.Status.Profile != "" {
+		// Integration already has a profile
+		return integration.Status.Profile
+	}
+	if p.Status.Profile != "" {
+		// Use platform profile if set
+		return p.Status.Profile
+	}
+	if knative.IsEnabledInNamespace(ctx, c, integration.Namespace) {
+		return v1.TraitProfileKnative
+	}
+	return platform.GetProfile(p)
 }

--- a/pkg/install/optional.go
+++ b/pkg/install/optional.go
@@ -2,6 +2,7 @@ package install
 
 import (
 	"context"
+
 	"github.com/apache/camel-k/pkg/client"
 	"github.com/go-logr/logr"
 )

--- a/pkg/trait/knative.go
+++ b/pkg/trait/knative.go
@@ -69,6 +69,8 @@ type knativeTrait struct {
 	// that can be removed in a future version of Knative, filtering is enabled only when the integration is
 	// listening from more than 1 channel.
 	FilterSourceChannels *bool `property:"filter-source-channels"`
+	// Enables Knative CamelSource pre 0.15 compatibility fixes (will be removed in future versions).
+	CamelSourceCompat *bool `property:"camel-source-compat"`
 	// Enable automatic discovery of all trait properties.
 	Auto *bool `property:"auto"`
 }
@@ -173,6 +175,30 @@ func (t *knativeTrait) Configure(e *Environment) (bool, error) {
 }
 
 func (t *knativeTrait) Apply(e *Environment) error {
+	// To be removed when Knative CamelSources < 0.15 will no longer be supported
+	// Older versions of Knative Sources use a loader rather than an interceptor
+	if t.CamelSourceCompat == nil || *t.CamelSourceCompat {
+		for i, s := range e.Integration.Spec.Sources {
+			if s.Loader == "knative-source" {
+				s.Loader = ""
+				util.StringSliceUniqueAdd(&s.Interceptors, "knative-source")
+				e.Integration.Spec.Sources[i] = s
+			}
+		}
+	}
+	// End of temporary code
+
+	if e.IntegrationInPhase(v1.IntegrationPhaseInitialization) {
+		// Interceptor may have been set by a Knative CamelSource
+		if util.StringSliceExists(e.getAllInterceptors(), "knative-source") {
+			// Adding required libraries for Camel sources
+			util.StringSliceUniqueAdd(&e.Integration.Status.Dependencies, "mvn:org.apache.camel.k/camel-knative")
+			util.StringSliceUniqueAdd(&e.Integration.Status.Dependencies, "mvn:org.apache.camel.k/camel-k-runtime-knative")
+			// Adding platform HTTP
+			util.StringSliceUniqueAdd(&e.Integration.Status.Capabilities, v1.CapabilityPlatformHTTP)
+		}
+	}
+
 	if len(t.ChannelSources) > 0 || len(t.EndpointSources) > 0 || len(t.EventSources) > 0 {
 		util.StringSliceUniqueAdd(&e.Integration.Status.Capabilities, v1.CapabilityPlatformHTTP)
 	}

--- a/pkg/trait/trait_types.go
+++ b/pkg/trait/trait_types.go
@@ -25,6 +25,7 @@ import (
 	"strconv"
 	"strings"
 
+	"github.com/apache/camel-k/pkg/util"
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	controller "sigs.k8s.io/controller-runtime/pkg/client"
@@ -735,4 +736,16 @@ func (e *Environment) getIntegrationContainer() *corev1.Container {
 	}
 
 	return e.Resources.GetContainerByName(containerName)
+}
+
+func (e *Environment) getAllInterceptors() []string {
+	res := make([]string, 0)
+	util.StringSliceUniqueConcat(&res, e.Interceptors)
+
+	if e.Integration != nil {
+		for _, s := range e.Integration.Sources() {
+			util.StringSliceUniqueConcat(&res, s.Interceptors)
+		}
+	}
+	return res
 }

--- a/pkg/util/util.go
+++ b/pkg/util/util.go
@@ -88,6 +88,16 @@ func StringSliceUniqueAdd(slice *[]string, item string) bool {
 	return true
 }
 
+// StringSliceUniqueConcat append all the items of the "items" slice if they are not already present in the slice
+func StringSliceUniqueConcat(slice *[]string, items []string) bool {
+	for _, item := range items {
+		if !StringSliceUniqueAdd(slice, item) {
+			return false
+		}
+	}
+	return true
+}
+
 // EncodeXML --
 func EncodeXML(content interface{}) ([]byte, error) {
 	w := &bytes.Buffer{}

